include/dali/core/mm/cuda_vm_resource.h

```diff
@@ -0,0 +1,616 @@
+      available_blocks--;
+      return mem;
+    }
+
+    void append(va_region &&other) {
+      assert(address_range.end() == other.address_range.ptr());
+      assert(block_size == other.block_size);
+      address_range.size() += other.address_range.size();
+      int other_blocks = other.num_blocks();
+      int old_blocks = num_blocks();
+      mapping.resize(old_blocks + other_blocks);
+      for (int i = 0; i < other_blocks; i++) {
+        mapping[old_blocks + i] = other.mapping[i];
+        other.mapping[i] = {};
+      }
+      other.mapping = {};
+
+      mapped.append(other.mapped);
+      available.append(other.available);
+
+      other.available = {};
+      other.mapped = {};
+      available_blocks += other.available_blocks;
+      other.available_blocks = 0;
+      assert(mapping.size() == mapped.size());
+      assert(mapping.size() == available.size());
+    }
+
+    void resize(int new_num_blocks) {
+      if (new_num_blocks < num_blocks()) {
+        int no_longer_in_range = 0;
+        for (int i = new_num_blocks; i < num_blocks(); i++) {
+          if (mapping[i]) {
+            cuvm::Unmap(block_ptr(i), block_size);
+            CUDA_DTOR_CALL(cuMemRelease(mapping[i]));
+            mapping[i] = {};
+            if (available[i])
+              no_longer_in_range++;
+          }
+        }
+        available_blocks -= no_longer_in_range;
+      }
+      mapping.resize(new_num_blocks);
+      mapped.resize(new_num_blocks);
+      available.resize(new_num_blocks);
+      address_range.size() = new_num_blocks * block_size;
+    }
+
+    void set_block_availability(char *start, char *end, bool availability) {
+      int blk_start = (start - block_ptr<char>(0)) / block_size;
+      int blk_end   = (end   - block_ptr<char>(0)) / block_size;
+      int flipped = 0;
+      for (int b = available.find(!availability, blk_start); b < blk_end; ) {
+        int first_flipped = available.find(availability, b+1);
+        flipped += first_flipped - b;
+        b = first_flipped + 1;
+      }
+      if (availability)
+        available_blocks += flipped;
+      else
+        available_blocks -= flipped;
+      available.fill(blk_start, blk_end, availability);
+    }
+
+    cuvm::CUAddressRange    address_range;
+    vector<mem_handle_t>    mapping;
+    bitmask                 mapped, available;
+    ptrdiff_t               available_blocks = 0;
+    size_t                  block_size = 0;
+  };
+
+  SmallVector<va_region, 8> va_regions_;
+  SmallVector<cuvm::CUMemAddressRange, 8> va_ranges_;
+  size_t initial_va_size_ = 0;
+  size_t block_size_ = 0;
+  size_t total_mem_ = 0;
+  int device_ordinal_ = -1;
+
+  void adjust_params(size_t &size, size_t &alignment) {
+    alignment = std::max(alignment, next_pow2(size >> 11));
+    alignment = std::min(alignment, block_size_);
+    size = align_up(size, alignment);
+  }
+
+  void *do_allocate(size_t size, size_t alignment) override {
+    adjust_params(size, alignment);
+    if (size == 0)
```

Comment:
Can we check for size as the first thing in the function?

Reply:
We can - not like it matters much, though.

Reply:
Done.